userlib.go

```diff
@@ -411,42 +588,167 @@ func symEnc(key []byte, iv []byte, plaintext []byte) []byte {
+	if SymbolicDebug {
+		record(plaintext, `{"userlib.SymDec": {"key": %s, "ciphertext": %s}}`, resolve(key), resolve(ciphertext))
+	}
 
 	return plaintext
 }
 
 var SymDec = symDec
+
+/*
+********************************************
+**         BETA: Symbolic Debugger        **
+********************************************
+
+The Symbolic Debugger is a debugging tool that
+we created to assist with debugging Datastore
+entries. Read the comments below to see what you
+can do with this.
+*/
+
+// This flag enables the Symbolic Debugger
+// If the Symbolic Debugger is causing excessive slowdown or memory usage,
+// you may want to set this to false.
+var SymbolicDebug = true
+
+// This flag enables verbose logging through the Symbolic Debugger
+// If you set this to true, you may want to pipe your output to a file
+// e.g. go test -v > debug.txt
+// to see the full, untrunctated debug logging
+var SymbolicVerbose = true
+
+// This flag sets the maximum key length in the debugger
+// You may want to increase this to a larger number (e.g. 100) if you're using
+// identifiers that have meaning, e.g. "username-{}-filename-{}", and want to see
+// the entire identifier in the symbolic representation.
+var SymbolicMaxLength = 5
+
+// Feel free to use userlib.DebugMsg(...) to print strings to the console.
+func DebugMsg(format string, args ...interface{}) {
+	msg := fmt.Sprintf("%v ", time.Now().Format("15:04:05.00000"))
+	log.Printf(msg+strings.Trim(format, "\r\n ")+"\n", args...)
+}
+
+// The Symbolic Debugger lets you export a snapshot of the Datastore in a symbolic representation.
+// To do this, call DebugExportDatastore("datastore.json"), or any file name of your
+// choice.
+func DebugExportDatastore(outputFile string) {
+	entries := make([]string, 0)
+	for key, element := range datastore {
+		entries = append(entries, fmt.Sprintf(`{"Key": %s, "Value": %s}`, resolve(key[:]), resolve(element)))
+	}
+	output := fmt.Sprintf(`[%s]`, strings.Join(entries, ","))
+	var prettyOutput bytes.Buffer
+	_ = json.Indent(&prettyOutput, []byte(output), "", "  ")
+	_ = ioutil.WriteFile(outputFile, prettyOutput.Bytes(), 0644)
+}
+
+// ----------- SYMBOLIC DEBUGGER: PRIVATE METHODS [BEGIN] ---------------
+
+// Takes in byte slice data (with no text interpretation) and cuts it short
+func truncateBytes(data []byte) (truncated string) {
+	return truncateStr(fmt.Sprintf("%x", data))
+}
+
+// Takes in string data (with text interpretation) and cuts it short
+func truncateStr(data string) (truncated string) {
+	if len(data) < SymbolicMaxLength+3 {
+		truncated = data
+	} else {
+		truncated = data[:SymbolicMaxLength] + "..."
+	}
+	data_bytes, err := json.Marshal(truncated)
+	if err != nil {
+		panic("symbolic debugger failed in truncateStr with error: " + err.Error())
+	}
+	return fmt.Sprintf("%s", data_bytes)
+}
+
+// Converts byte slice data to valid symbolic map keys.
+func keyFromBytes(data []byte) (truncated string) {
```

Comment:
Suggestion to improve readability: rename this function to distinguish from an encryption key.
```
// Converts byte slice data to a fixed size value that is suitable to use as the
//storage key in a map.

func mapKeyFromBytes(data []byte) (truncated string) {
	return fmt.Sprintf("%x", sha512.Sum512(data))
}
```